### ISSUE TITLE
kedro-datasets: Rename MatplotlibWriter to MatplotlibDataset 

### ISSUE THREAD

**Issue by Galileo-Galilei** [kedro-org/kedro-plugins#353]
## Description
`MatplotlibWriter` has been the only dataset with inconsistent naming for a while. There are a couple of more subtle issues with it still to be fixed (see 
https://github.com/kedro-org/kedro-plugins/issues/529), but I think we should use the dataset renaming (which is a breaking change whatever) as an opportunity to make it consistent with the rest of the codebase

## Context

Consistency is better :)

## Possible Implementation
Rename `MatplotlibWriter` to `MatplotlibDataset`.

## Possible Alternatives
 Don't do it ;)

This is likely a great ticket for Hacktoberfest if the project is eligible?

**Comment by noklam** [comment]
Nice idea, I asked in the team and start tagging issues.

**Comment by astrojuanlu** [comment]
1. This affects Kedro-Viz, since they use `MatplotlibWriter` in various places. To retain backwards compatibility, they would probably have to do some `try ... except` with imports. cc @rashidakanchwala @tynandebold 
2. I think this goes to some of our findings in https://github.com/kedro-org/kedro/issues/1936: that we don't really intend to implement `_load` for certain datasets (including this one) because they're meant to be _artifacts_ and not really full I/O components.

Regardless, for consistency I agree we should rename it 👍🏽

**Comment by noklam** [comment]
https://kedro-org.slack.com/archives/C03RKP2LW64/p1695929737315389

An user ask why `load` is not supported? Maybe this is the reason why it is called Writer but not Dataset. I don't use this dataset myself so I can't remember is it always like this.

**Comment by astrojuanlu** [comment]
Maybe we could have a `MatplotlibFigureDataset` (that saves and loads figures objects as pickles) (although maybe we shouldn't and we should tell users to use `PickleDataset` for this) and `MatplotlibImageDataset` (that only saves images and has no `_load`)

**Comment by Galileo-Galilei** [comment]
Actually it totally makes sense that some dataset don't have ``_save`` method (API ``Dataset`` didn't for a while, ``SQLQueryDataset`` still doesn't...). Can I start a PR for this or the impact is too high on kedro-viz?

**Comment by astrojuanlu** [comment]
Yeah I don't think it's a problem that some datasets don't have `_save`. What's problematic is the monkeypatching that kedro-viz does on some of them https://github.com/kedro-org/kedro-viz/issues/1352 

About the impact on Kedro Viz, cc @rashidakanchwala @tynandebold

**Comment by rashidakanchwala** [comment]
Yes, it will. We will be looking at making sure Kedro-viz is not so coupled with Kedro-datasets. And we are going to prioritise this work in the second half of November most likely. Until then, we could pin kedro-viz to kedro-datasets to the current version.

**Comment by astrojuanlu** [comment]
I don't think this is a good first issue, there's some uncertainty still https://github.com/kedro-org/kedro-plugins/issues/353#issuecomment-1740974418 removing the label, when we have more clarity we'll update this issue

**Comment by merelcht** [comment]
Kedro-Viz has made significant changes to decouple from `kedro-dataset`, would renaming `MatplotlibWriter` still be an issue @rashidakanchwala ?

**Comment by rashidakanchwala** [comment]
nope this is not an issue. Kedro-viz only reads NewType = 'image' from preview method so it should be fine.
